src/tools/pkcs11-tool.c

```diff
@@ -4104,6 +4104,16 @@ static int test_signature(CK_SESSION_HANDLE sess)
 		modLenBits = get_private_key_length(sess, privKeyObject);
 		modLenBytes = (modLenBits + 7) / 8;
 
+		/* Fill in data[0] and dataLens[0] */
+		dataLen = modLenBytes;
+		data[0] = 0x00;
+		data[1] = 0x01;
+		memset(data + 2, 0xFF, dataLen - 3 - dataLens[1]);
+		data[dataLen - 36] = 0x00;
+		memcpy(data + (dataLen - dataLens[1]), datas[1], dataLens[1]);
+		datas[0] = data;
+		dataLens[0] = dataLen;
```

Comment:
Except for maybe data[0] and data[1], the duplicated code is necessary: In each run through the while loop, the key is different, maybe with different modLenBits, as is the case with my 4096/../4096/1792 keyset (the unchanged v.0.16.0 code ran into the error, that cardxy_compute_signature got passed a wrong in_length of 512 instead of 224 bytes for the 1792 bit RSA key and wrong in_data).
A better place for data[0] = 0x00; and //data[1] = 0x01; would be before the while loop: I couldn't find where data[0] is set elsewhere after the (new, faultily placed) call to pseudo_randomize in git/master (but if You are shure it has value 0x00 when entering this test, I propose it to end up as comment). Setting data[1] = 0x01 is a clear duplication from 3.test and would be better off a comment. My reasoning was/is to have the complete content of data for the 4. test visible in this place (not scattered among a couple of tests; maybe 3.test's data setting changes sometime?). Thus my new proposal resumes to:

	j = 1;  /* j-th signature key */
	data[0] = 0x00; // maybe line commented out, if data[0] has already 0x00 value
//	data[1] = 0x01;
	while (find_object(sess, CKO_PRIVATE_KEY, &privKeyObject, NULL, 0, j++) != 0) {
		CK_ULONG	modLenBits;

		label = getLABEL(sess, privKeyObject, NULL);
		modLenBits = get_private_key_length(sess, privKeyObject);
		modLenBytes = (modLenBits + 7) / 8;

		/* Fill in datas[0] and dataLens[0] */
		dataLen = modLenBytes;
		memset(data + 2, 0xFF, dataLen - 3 - dataLens[1]);
		data[dataLen - 36] = 0x00;
		memcpy(data + (dataLen - dataLens[1]), datas[1], dataLens[1]);
		datas[0] = data;
		dataLens[0] = dataLen;